### PR TITLE
acgan: Add batch normalization to the Generator, etc

### PR DESCRIPTION
**Pull request by ozabluda**
1. Add batch normalization to the Generator. This makes the example closer to the referenced paper and improves generated images. Adding batch normalization to the Discriminator breaks training so badly, that I suspect a bug (maybe https://github.com/fchollet/keras/pull/5647 is fixed incompletely or something). Not adding batch normalization to the Discriminator also side-steps the issue of correlation of samples within a batch (https://github.com/soumith/ganhacks#4-batchnorm)

2.  Use one-sided soft labels and a harder `soft_one=0.95` vs 0.9). The referenced paper says they don't need one-sided soft labels. This example also doesn't "need" them any more, but the generated images are better. Add reference to a paper.

3. Increase `epoch=100` from 50, as good images often appear between epochs 50 and 100. Note that the training time per epoch is half that of the original example, after https://github.com/fchollet/keras/pull/8482/commits/67cd3b09510643655998311d66f0e9b3475f32e5

4. Increase output precision of various losses from 2 decimal digits to 4. You can't really tell what is going on with just 2. 

@lukedeo , I see a lot of examples online which use embedding with Hadamard, but do you know of any paper(s) we can reference? I haven't seen it in any of the GAN papers. I really like embedding with Hadamard, as replacing them would require multiple (3-5?) additional layers, but to be thorough I did a half-hearted attempts to remove them (make closer to the acgan paper), just to see if I can, and failed (generated images are much worse).

### CODE REVIEW

**Comment by ozabluda** [comment]
Generated images are better than previous best (https://github.com/fchollet/keras/pull/8482#issuecomment-344401396)

The biggest difference is much larger diversity in line thickness.

Epoch 57
plot_epoch_057_generated

Epoch 99
plot_epoch_099_generated

Epoch 100
plot_epoch_100_generated

**Comment by fchollet** [comment]
Note that if you want to completely freeze a model that has BN layers (like here) you need to do more than just set it to non-trainable, you need to also disable its updates. Otherwise the batch statistics will still get updated during training. This is not a bug, this is simply a manifestation of the fact that non-backprop updates and layer trainability are independent (for instance, if you set a stateful RNN to non-trainable, that will not freeze its state).

It is possible that we should have a `frozen` attribute that would both disable trainability and other updates. Or some other API to freeze BN layers.

**Comment by alpapado** [comment]
> you need to also disable its updates.

Could you elaborate on that please? Do you mean setting every layer's trainable attribute to False?

**Comment by fchollet** [comment]
No. You'll need to clear the attribute `_per_input_updates` on every sublayer. That, or find other way to get `model.updates` to return `[]`.

**Comment by ozabluda** [comment]
I don't think it ever makes sense to freeze those 2 _trainable_ parameters, while updating the 2 _non-trainable_ parameters. This approach should properly be called "Breaking Deep Network Training by Introducing Uncorrectable Internal Covariate Shift". In terms of calling things _trainable_ or _non-trainable_, I don't see what difference does it make conceptually whether the parameter updates happen during forward prop or backprop. I think those _non-trainable_ weights should always be frozen as well, and renamed _trainable_when they aren't frozen.

This issue does not affect this PR (which has BN only on the Generator, which is never frozen), but may be the reason why my attempts to put BN into the Discriminator broke things very badly. The following workaround https://github.com/fchollet/keras/issues/4762#issuecomment-321299663 seems to be working for the GAN use case, and I'll try it, but not for a while.

**Comment by fchollet** [comment]
@ozabluda Jeremy Howard did a study on this recently ("should we update the batch statistics during fine-tuning?") and the answer was, "it depends". It's not clear-cut.

Beyond batch norm specifically, you seem confused by the different between trainability and stateful behavior. Setting a layer to non-trainable means its trainable weights will not be taken into account during training. It does not affect the parts of the layer's state that are independent from training. For instance, a layer that maintains a counter that is incremented by one with every batch, will not stop doing that if you set `trainable = False`. Because that has nothing to do with training. Same with stateful RNNs, or with BN updates.

If you want to run your BN layers in *inference mode*, the way to do it is to pass a static boolean as the `training` argument:

```python
y = BatchNormalization()(x, training=False)
```

**Comment by lukedeo** [comment]
unrelated, @ozabluda I don't know of any references. Tbh, it always just seemed more natural on an intuitive basis.

**Comment by ozabluda** [comment]
@fchollet, Can't find that study by Jeremy Howard.

Is there a recommended way to freeze those two _non_trainable_ weights that would work for the acgan example, if I put BatchNormalization into the Discriminator? `y = BatchNormalization()(x, training=False)` will work for the combined model, but not for the Discriminator. For the same reason, the idea from my previous comment will not work either.

Right now, the best advice seems to be to go through the discriminator model, immediately after `discriminator.trainable = False` is set, looking for BatchNormalization layers and set their `per_input_updates={}`. Is that supposed to work without screwing up the already compiled discriminator?

Also see https://github.com/tensorflow/tensorflow/issues/10857

**Comment by ahundt** [comment]
> It is possible that we should have a frozen attribute that would both disable trainability and other updates. Or some other API to freeze BN layers.

Yes, please!

**Comment by fchollet** [comment]
> Is there a recommended way to freeze those two non_trainable weights that would work for the acgan example

There is no "clean" way at the moment, but we need one. The simplest would be a layer/model attribute that regulates whether or not the layer/model will always return `[]` when asked for `.updates`. It would be very straightforward to implement. Could be named `frozen`, `freeze`, `freeze_updates`...

**Comment by ozabluda** [comment]
FWIW, none of the following works to freeze those 2 weights in BatchNormalization I put into Discriminator in acgan example:

Immediately after `discriminator.trainable = False`
```
    for layer in discriminator.layers[1].layers:
        print(layer.name)
        if layer.name.startswith('batch_normalization_'):
            layer._per_input_updates={}
```

before and after `combined.compile()`:
```
    for layer in combined.layers[-1].layers[1].layers:
        if layer.name.startswith('batch_normalization_'):
            print(layer.name)
            layer._per_input_updates={}
```

**Comment by ahundt** [comment]
I agree, `freeze_updates` sounds like the best of your suggestions

**Comment by ozabluda** [comment]
@ahundt, I have no immediate plans to work on it, since I don't understand that part of the Keras code, and therefore the suggestions (as of now). I also don't understand the performance implications, for example compared to native https://github.com/tensorflow/tensorflow/pull/12580.

So if you feel like doing it, it would be great. For the reference, this is how I check if freezing actually works in the acgan example after putting BatchNormalization into the Discriminator, (not in this PR), immediately after `combined.train_on_batch()`
```
print(combined.layers[-1].layers[1].layers[5].name)
print(np.all([
    np.array_equal(w1, w2) for w1,w2 in
    zip(weights,
        combined.layers[-1].layers[1].layers[5].get_weights())]))

print(np.array_equal(
    weights[2],
    combined.layers[-1].layers[1].layers[5].get_weights()[2]))#0-1 are equal, 2-3 are not
```
A much simpler example can't be made, because this is the use case for which none of the hacks, etc work.

**Comment by ahundt** [comment]
@ozabluda sorry, meant to @ mention fchollet on that previous post, I was addressing the feature he was suggesting.

**Comment by ahundt** [comment]
@fchollet Also, I'm not sure I mentioned this, but freezing batch normalization updates is quite helpful when fine tuning segmentation problems from pre-trained weights. I think something like this [modified BatchNormalization class with freeze](https://github.com/broadinstitute/keras-resnet/blob/master/keras_resnet/layers/_batch_normalization.py) would do the trick. Would it be acceptable to place the `freeze` parameter in `allowed_kwargs` under `Layer`?

**Comment by fchollet** [comment]
For the record as of master you can set `layer.updatable = False` or `model.updatable = False` to freeze updates (mirrors `trainable = False`).

@lukedeo would you like to review this PR?

**Comment by lukedeo** [comment]
yep, can check by EOW

**Comment by fchollet** [comment]
@ahundt @ozabluda FYI I think the `updatable` system is overcomplicated. Pragmatically, what most people want/expect is that setting `bn.trainable = False` will run BN in inference mode, i.e. without updates. Likewise, pragmatically, updates when `trainable == False` only make sense for stateful layers (stateful RNNs being the only example of that).

Thus I am reverting `updatable` (it was never part of a release) and simply disabling updates for non-stateful layers when `trainable == False`.

**Comment by fchollet** [comment]
@lukedeo if you're still available to review this PR, we're waiting for your input. Otherwise, please tell and we will find another reviewer. Thanks!

**Comment by ahundt** [comment]
@fchollet thanks for the update, the new functionality you describe is the behavior I imagined when I first saw the name of the flag way back as I was learning the code base. I look forward to using it!

**Comment by ozabluda** [comment]
For the reference, the changes described in https://github.com/keras-team/keras/pull/8616#issuecomment-357029549 were made in https://github.com/keras-team/keras/commit/24246ea53157f7e9faef74db3e4c06642a50509d

**Comment by ahundt** [comment]
Also the reverse commit that makes trainable=False prevent updates is https://github.com/keras-team/keras/commit/24246ea53157f7e9faef74db3e4c06642a50509d

**Comment by ahundt** [comment]
Specifically, the reverse commit that makes `trainable=False` prevent updates is  https://github.com/keras-team/keras/commit/24246ea53157f7e9faef74db3e4c06642a50509d

**Comment by lukedeo** [comment]
@fchollet this seems fine to me overall. 

@ozabluda one question, though not necessarily needed for this PR, is if we should also add BN to the discriminator since it's technically closer to the paper. If we see good performance, which we do, we don't necessarily need it, but just raising as people might question now that I think about it. The one place I could see this *maybe* helping is if people want to adapt this script to CIFAR10 or that ilk.

**Comment by ozabluda** [comment]
I tried adding BN to the discriminator (the very first message in this PR, item 1). With recent fixes to the BN, maybe I'll try it again later. An attempt to adapt it to CIFAR-10 is going on here: https://github.com/keras-team/keras/pull/8937

**Comment by ozabluda** [comment]
I've created animations of acgan training:
https://www.youtube.com/playlist?list=PL7zaNUNu3zI8O6cBK6rgpLCEXtZerDzU_

Resolutions are 1080p (2K), 1440p (3K), 2160p (4K) 7680p (8K). There are two types of video
acgan1: one frame per epoch.
acgan0: one frame per iteration.

Note that on YouTube, you can change the playback speed, and, when paused, go frame-by-frame with ',' and '.'

Real images are on the bottom. Fake generated one are on top. Epoch/iteration is in the gray bar in between. For every column of 10 digits (0-9), latent noise vector is the same (class is different). For every row of same digits, latent noise vector is different (class is the same).

Discriminator probability real/fake is shown as a grayscale square around a digit. It's scaled such that p=0.5 is pure black, p=0.75 is pure white, with p<0.5 and p>0.75 clipped to min and max values. Otherwise, one wouldn't be able to see what is going on.

Staring at these videos for a while, falsified a lot of my preconceived notions of what GAN is actually doing. Too many to describe here. I highly recommend anyone who is interested to stare for a while as well.

**Comment by ahundt** [comment]
@ozabluda can you fix the youtube link? doesn't seem to work for me

**Comment by ozabluda** [comment]
@ahundt, fixed.

**Comment by ahundt** [comment]
thanks! looks neat

**Comment by ozabluda** [comment]
>Staring at these videos for a while, falsified a lot of my preconceived notions of what GAN is actually doing. Too many to describe here. I highly recommend anyone who is interested to stare for a while as well.

Two quick things:
1. In this GAN, diversity doesn't come from training (after first epochs). Most It comes from sampling latent noise. Which is not what we want, but may not matter much in practice, depending.
2. I blame everything on the discriminator now, unlike the consensus, and what I though before I started at these videos for a while.

P.S. I've added acgan0 8K video - 30 min, 100 GB, baby!

**Comment by ahundt** [comment]
> P.S. I've added acgan0 8K video - 30 min, 100 GB, baby!

I present you with: 

🥇 

Awarded for the highest resolution 28x28 image.

P.S. Any hints on what rendering tools you used to generate these? Sounds like something useful

**Comment by ozabluda** [comment]
just ffmpeg. the code to generate pngs is an embarrassing mess, not sure if it's worthwhile to clean and add it to the example.

**Comment by ahundt** [comment]
A png utility could go in some non-keras repo or keras-contrib.

**Comment by Evan05543071** [comment]
> Specifically, the reverse commit that makes `trainable=False` prevent updates is [24246ea](https://github.com/keras-team/keras/commit/24246ea53157f7e9faef74db3e4c06642a50509d)

Does this mean:
I found that if I set the trainable= False (batch_normalization)
All parameters in batch normalization layer would stop updating
However,if I use what you said up there set the training =False
The moving mean and moving variance would be fixed and the others would still change while training
